Automattic/wp-calypso
# Experiment: load gridicons from external SVG file.

Gridicons are currently being wrapped in JSX and included as a separate JS bundle. This is suboptimal for a number of reasons, among which:
- JS is byte-for-byte much more CPU-demanding than SVG, a concern on mobile devices. The fewer JS bytes the browser needs to handle, the better.
- Encoding image data as JS invalidates any optimizations the browser may otherwise be able to perform, such as download prioritization.

This PR experiments with an alternate approach: SVG external content.

SVG data is kept in a static file, with each icon being assigned an ID.
We then have a small component which simply references the correct ID in this external file (using `<use>`, a native SVG feature), leaving it to the browser to fetch the data at an appropriate time during bootup.

Since the SVG data is all kept in a single file, we strike a good balance between cacheability and number of requests, we avoid having large amounts of procedurally-generated JSX, and we delegate all image fetching to the browser, which can apply the usual optimizations as it sees fit.

Potential caveats that were tested and appear to be handled correctly in this PR:
- IE11 and older versions of modern browsers work correctly, through the `svg4everybody` polyfill (~1KB). The polyfill is explicitly excluded from the evergreen bundle, as it's not needed.
- Icons are aligned correctly by copying existing alignment logic. Some minor changes to existing CSS rules were needed.
- A blank icon is shown when the icon name is unknown.

**Note**: this PR is not intended to be merged as-is. I'm looking to validate the approach before moving forward. If it's approved, the work will likely need to be split between this repo and the `gridicons` repo, with updates to the `gridicons` NPM package.

#### Changes proposed in this Pull Request

* Remove `AsyncGridicons` component
* Add `ExternalGridicons` component
* Copy alignment data from `gridicons` repo and use in `ExternalGridicons`
* Add tiled SVG file to `static` (formerly `public`), copied from the `gridicons` repo.
* Modify gridicon CSS rules to be compatible with this approach
* Add `svg4everybody` polyfill to support SVG external content in old browsers

#### Testing instructions

* Open the `calypso.live` branch of this PR in one browser tab.
* Open production in another browser tab.
* Alternate between both tabs, comparing as many different Calypso pages as you see fit. They should all be identical, pixel for pixel.
